### PR TITLE
Staging/capi selftest

### PR DESCRIPTION
**Pull request by cristina-suteu**
## Pull Request Description

  This PR implements a platform-agnostic system test of the CAPI platform drivers.

  Adds the CAPI selftest project, a comprehensive integration suite for the Common API platform drivers across Xilinx (Zynq-7000) and STM32 (NUCLEO-F767ZI) targets.

  The selftest validates CAPI peripherals through a type-agnostic test framework that remains inert until each driver is mapped on the target platform. Every peripheral can be independently gated via Kconfig
  (`CONFIG_CAPI_*`); tests compile only when their dependencies are met, and skip gracefully at runtime when a peripheral is absent.

  ### Examples

  - **basic_example**: Minimal UART transmit sanity check. Entry point for new users and CI smoke tests.
  - **capi_loopback**: Comprehensive suite exercising GPIO, IRQ, SPI, Timer, I2C, DMA, and UART through real hardware loopback and the test framework. Selectable per-peripheral skip modes for partial runs.

  ### Test Framework

  - Structured result tracking (pass / fail / skip with per-subtest granularity).
  - Watchdog hooks for cleanup on failure, configurable per-section.
  - Console UART transport for all output, with chunked polled writes to avoid TX FIFO stalls.

  ### Peripherals Tested

  Each test exercises loopback, data integrity, timing, and async delivery (IRQ or DMA where supported):

  - **GPIO**: Port and pin loopback over a wired pair (PS EMIO on Xilinx, PE0↔PC0 on STM32).
  - **IRQ**: GPIO-driven integration path validation (arm / ack / disarm hooks).
  - **SPI**: Clock modes, LSB-first, transfer sizes, and async abort.
  - **Timer**: Counter movement, compare/capture, overflow IRQ, rate calibration.
  - **I2C**: Two-bus setup (initiator + target on cross-wired SCL/SDA); address reconfig, repeated-start, NACK handling, and async target IRQ.
  - **DMA**: Memory-to-memory transfers (STM32 only).
  - **UART**: Async loopback (second instance, never console) on PS/PL backends on Xilinx.

  ### Platform Mappings

  **Xilinx:**
  - PS and PL variants coexist where available (gpio_ps + gpio_pl, uart_ps + uart_pl_lite + uart_pl_ns550, etc.), selected at compile time via `XPAR_*_NUM_INSTANCES`.
  - IRQ controller is auto-detected (GIC vs. AXI INTC) or manually selected.
  - Per-peripheral async support gated on IRQ availability.

  **STM32:**
  - Single instance per peripheral (NUCLEO-F767ZI pinout).
  - I2C2 brought up by a platform hook since CubeMX only generates I2C1.
  - Async tests run over the NVIC.

  ### Build System

  - Root `CMakeLists.txt` / `Kconfig`: detects `CONFIG_CAPI=y` in project defconfig and selects `capi/platform/${PLATFORM}/toolchain.cmake` for CAPI projects.
  - `capi/CMakeLists.txt` / `capi/Kconfig`: compiles CAPI dispatchers and platform-specific backends via per-module `CONFIG_CAPI_*` gates (UART, GPIO, IRQ, SPI, Timer, I2C, DMA), all default off.
  - `capi/platform/{xilinx,stm32}/CMakeLists.txt`: lists PS and PL variant sources; each source guards its body on `XPAR_*_NUM_INSTANCES` (Xilinx) so only instantiated peripherals compile.
  - `projects/capi_selftest/`: selectable board configs (`basic.conf` / `loopback.conf`) with per-board overrides (coraz7s / nucleo-f767zi) for loopback wiring and peripheral selection.

  ### Testing

  Validated on:

  - **Xilinx:** Cora Z7 (Zynq-7000) with GIC-only for PS/PL peripheral loopback.
  - **STM32:** NUCLEO-F767ZI across GPIO, SPI, I2C, DMA, and UART loopback paths.

  Testing covered initialization, functional loopback, interrupt delivery, and deinitialization with register readback.

  ## PR Type

  - [x] New feature (change that adds new functionality)
  - [ ] Bug fix (change that fixes an issue)
  - [ ] Breaking change (has dependencies in other repos or will cause CI to fail)

  ## PR Checklist

  - [x] I have followed the Coding style guidelines
  - [x] I have complied with the Submission Checklist
  - [x] I have performed a self-review of the changes
  - [x] I have commented my code, at least hard-to-understand parts
  - [x] I have built all projects affected by the changes in this PR
  - [x] I have tested in hardware affected projects, at the relevant boards
  - [x] I have signed off all commits from this PR
  - [x] I have updated the documentation (wiki pages, ReadMe etc), if applies

### CODE REVIEW

**Comment by adi-ramonimbao** [comment]
Hi. I can't get the GPIO port loopback test to pass on the MAX32657. Since there's only one GPIO port, I don't know how to set it up such that P0.0 and P0.1 are the pin pair to be tested.

I've written a GPIO pin loopback test that should not need any additional changes to the STM32 and Xilinx platforms:
https://github.com/adi-ramonimbao/no-OS/blob/staging/capi-selftest-max32657/projects/capi_selftest/src/examples/capi_loopback/tests/test_gpio.c#L102-L193

There are a few additions in `common_data.c` and `common_data.h` as well:

- https://github.com/adi-ramonimbao/no-OS/blob/staging/capi-selftest-max32657/projects/capi_selftest/src/common/common_data.c#L126-L152
- https://github.com/adi-ramonimbao/no-OS/blob/staging/capi-selftest-max32657/projects/capi_selftest/src/common/common_data.h#L29-L56

**Comment by buha** [comment]
writing it down for tracking, wip:

- [ ] cmake build, as main branch doesn't have make builds anymore
- [ ] remove unit tests
- [ ] add more integration tests like gpio-gpio interrupt, timer interrupt, spi/i2c interrupt

**Comment by Voc94** [comment]
## Changelog V1

  1. CMake Build System Transition
     - Migrate to CMake/Kconfig with board-specific configs (basic.conf, loopback.conf)

  2. GPIO Pin Loopback Test (per @adi-ramonimbao)
     - Add pin-level loopback (P0.0 output, P0.1 input on same port) for single-port
       boards (e.g., MAX32657)

  3. Integration Test Suite
     - Remove unit tests
     - Add GPIO loopback test: port-level and pin-level variants, direction control,
       high/low patterns
     - Add IRQ test: GPIO edge delivery, callback argument passing, enable/disable
       gating, reconnect, repeated edge handling
     - Add SPI loopback test: polled and async modes, interrupt callbacks, command
       verification(cs gpio out - capi_gpio in to verify)
     - Add timer test: three backends (TTC/PS, AXI/PL, SCU/PS) with overflow IRQ,
       rate counter, compare mode
     - Add I2C loopback test: initiator/target roles, polled and async modes,
       runtime bus speed configuration
     - Add DMA test: async transfers with scatter-gather and completion callbacks
     - Test framework: table-driven execution, skip categories (HW_ABSENT,
       IRQ_ABSENT, BSP_ABSENT, etc.), per-subtest skip flag, UART console output

  4. Platform Mappings
     - Xilinx: GPIO/IRQ/SPI/Timer/I2C backends with PS/PL selectors and auto-detect
     - STM32: GPIO pin pair
     - Platform hooks: GPIO-IRQ arm/ack/disarm (vendor-specific setup behind
       platform interface)
       
CI ERRORS: stem from capi-xilinx being outdated, will be fixed when we rebase to main after capi-xilinx merge

**Comment by Voc94** [comment]
## Changelog v2

### `src/platform/xilinx/parameters.h` (+12)

- Added `SPI_SEL_PL` above the auto-detection guard to select the PL SPI backend on Cora Z7S.
- `SPI_DEVICE_SPEED_HZ` is now `0U` for PL SPI because the AXI Quad SPI clock rate is fixed in HDL using `C_SCK_RATIO`. Passing a non-zero rate returns `-ENOTSUP`.

### `src/examples/capi_loopback/tests/test_gpio.c` (-16)

- Removed the direction readback checks from `gpio_loopback()` and `gpio_pin_loopback()`.
- Removed the calls to `capi_gpio_port_get_direction()` and `capi_gpio_pin_get_direction()`.
- Removed the related direction variables.

### `src/examples/capi_loopback/tests/test_i2c.c` (-37)

- Simplified `i2c_bus_speed()` to initialize I2C, run one standard-rate `i2c_xfer_check()`, and deinitialize I2C.
- Removed the calls that changed the bus speed to `I2C_SPEED_ALT` and then restored the standard speed.
- Moved `#endif /* I2C_TARGET_OPS */` above `i2c_bus_speed()`.

**Comment by Voc94** [comment]
## Changelog V3

cmake
- Root CMakeLists.txt: CAPI toolchain selector generalized from xilinx-only to capi/platform/${PLATFORM}/toolchain.cmake.
- Added STM32 CAPI toolchain: toolchain.cmake, FindSTM32CubeMX.cmake, FindSTM32CubeIDE.cmake. (needs cristina's approaval/edits)

per file
- parameters.h: default board now Cora Z7S (GPIO_SEL_PS/IRQ_SEL_GIC/SPI_SEL_PL/TIMER_SEL_TTC/I2C_SEL_PL); ZedBoard comments removed; timer macros fixed (XPAR_XTMRCTR_*, SCU clock = CPU/2); UART extra config added; committed merge-conflict markers removed across the series.
- basic.conf: trimmed to UART/ALLOC/TIME/IRQ + basic example.
- boards/: zed.conf→coraz7s.conf, added nucleo-f767zi.conf.
- nucleo-f767zi.ioc: new, UART-only (USART3 PD8/PD9).
- PL SPI paramters.h mappings for clk_freq fixed.
- test_gpio.c : put  capi_gpio_port_get_direction() and capi_gpio_pin_get_direction() tests back.
- test_i2c.c: added target-TX (sizes 1/16/32) and master-async coverage, now also checks the old address stops responding.
- test_framework.h: _OR_CLEANUP assertion forms.
- test_gpio/irq/spi/i2c/timer.c: irq usage in tests converted to _OR_CLEANUP .

**Comment by buha** [comment]
https://github.com/analogdevicesinc/no-OS/blob/staging/capi-selftest/projects/capi_selftest/src/platform/xilinx/main.c#L162

this #ifdef looks like a leftover, can you check please ? you need irq for i2c as well, not just for spi

**Comment by buha** [comment]
https://github.com/analogdevicesinc/no-OS/blob/staging/capi-selftest/capi/platform/xilinx/xilinx_capi_i2c_pl.c#L428

i think this should be a hard error in the capi implementation

**Comment by buha** [comment]
i2c_bus_speed doesn't change buss speed, i think it'd be worth at least testing that the speed changes in the expected direction, it's impossible to measure precisely, but if you measure a synchronous function execution time with different bus speeds you'll see a difference

**Comment by Voc94** [comment]
Changelog v4
- Rebased onto current capi-xilinx.
- xilinx/parameters.h
   - I2C IRQ support is now detected separately for master and target from the BSP.
   - Fixed fabric IRQ detection when the BSP has no fabric interrupts.
   - IRQ controller selection now uses IRQ_SEL_CASCADE.
   - Moved peripheral selection into one block.
   - moved UART_EXTRA_TYPE and UART_EXTRA_INIT to the right commit.
- common_data.h / common_data.c
   - I2C async tests now run only when both I2C roles are mapped and have IRQ support.
   - Added bus-speed timing limits and disabled unsupported duty-cycle testing on Xilinx.
   - irq_config is no longer tied to SPI.
   - irq_config now passes IRQ_CTRL_EXTRA, fixing cascaded IRQ setup.
- xilinx/main.c
   - The IRQ controller is initialized whenever IRQ_CTRL_IDENTIFIER exists, even in designs without SPI.
- test_i2c.c
   - Added on-wire I2C speed tests for slow/fast speed, duty cycle, and timing direction.
   - Increased the test payload from 4 to 1024 bytes so speed changes can be measured.
   - The fast transfer must be at least 10% faster than the slow transfer(verification against the uptimes between tests).
   - Tests skip when timing is too short to measure or a speed is unsupported.
   - The bus is always recovered after each speed probe.

**Comment by Voc94** [comment]
## Changelog v5

  test_uart

  - New UART external-loopback suite (test_uart.c/.h, wired into capi_loopback + CMake).
    One controller, TX strapped to RX; internal loopback never used. Stubs to
    NOT_CONFIGURED when UART_ASYNC_OPS is undefined.
  - Cases: BASIC (blocking TX/RX), ASYNC_BASIC (IRQ TX+RX, plus prebuffered / TX-busy
    -EBUSY / RX-timeout-continuation), ASYNC_SPEED (9600 vs 115200, fast leg must be
    \>=10\% shorter), ASYNC_IRQ (per-source masks + status predicates).
  - Every async case gated on capability axes (HAS_IRQ / HAS_LINE_CONFIG / HAS_IRQ_CTL
    / HAS_RX_TIMEOUT); a missing feature SKIPs, never FAILs.

  common_data

  - uart_async_config: second UART for the tests, separate from the console (the speed
    case reprograms the line rate mid-run, which would silence the log).

  xilinx parameters.h

  - Board config keyed to IRQ mode (GIC / CASCADE / no-IRQ), auto-detected from the BSP,
    each picking a different backend set off one file: GIC=PS/PL/PL/PS/TTC/NS550,
    CASCADE=PL/PS/PS/PL/AXI/Lite, no-IRQ=PS/PL/PL/PS/SCU/PS.
  - UART_ASYNC_* block: NS550/Lite/PS selection, per-core IRQ id encoding, HAS_IRQ
    derived from the same XPAR_*_INTERRUPTS macro that sets use_irq so async cases SKIP
    on a polled build. SPI falls back to use_irq=false when its INTERRUPTS macro is absent.

**Comment by Voc94** [comment]
Changelog v6

  Rebased onto origin/main now that staging/capi-xilinx is merged

  GPIO IRQ

  - fx_raise_edge() now drives a full low->high->low pulse instead of leaving the pin high. Every pulse costs the same number of input transitions this way; leaving it high made the first pulse one transition
  cheaper than the rest, so a change-triggered input saw 2N-1 events for N pulses -- which reads exactly like duplicate delivery.
  - New GPIO_IRQ_EVENTS_PER_EDGE: 1 for a rising-edge detector (PS XGpioPs, armed EDGE_RISING), 2 for a change-triggered one (PL AXI GPIO fires ip2intc_irpt on any input change, no edge-type control). Default
  1U in common_data.h, PL override in parameters.h.
  - irq_repeated_edges() waits for before + per_edge each iteration and asserts count == edges * per_edge. Adds a per_edge TEST_VALUE.

  Timer

  - AXI TIMER_COUNTER_MAX 0xFFFFFFFF -> 0x01000000. The driver preloads TLR = 2^32 - max, so a full span is ~42.9 s per overflow at 100 MHz and ASYNC_IRQ (1 s wait) saw zero deliveries. 0x01000000 is ~168 ms inside the timeout, still a wide 32-bit rollover, and well above the rate window.
  - SCU TIMER_INPUT_CLK_HZ now derives from XPAR_CPU_CORE_CLOCK_FREQ_HZ / 2 when the SDT BSP emits no XPAR_XSCUTIMER_0_CLOCK_FREQ (Vitis 2025.x gives the SCU timer only base addr + interrupt). Prefers a BSP
  value if a future toolchain publishes one. SCU TIMER_OUTPUT_FREQ_HZ 1000U -> 0U -- free-run unprescaled, the 8-bit prescaler can't reach 1 kHz from ~325 MHz.

  UART async

  - UART_ASYNC_LEN moved from test_uart.c into parameters.h (128U), with an #ifndef default in common_data.h. It's a per-backend size must exceed the deepest TX FIFO or the transfer drains during the fill
  and TX_BUSY has nothing to reject. The PS FIFO is 64, which is why the old tx[64] completed synchronously.


Needs Cristina's approval/edit for the uart & common_data changes for the STM32 build, then I think it's done.

**Comment by rbolboac** [comment]
please rebase on main and check if build is still failing

**Comment by Voc94** [comment]
###  Changelog v7

  Rebased onto current main.

  README.rst

  - Expanded the README into a full guide: what the suite proves (real MOSI→MISO loopback, real I2C initiator→target,
  etc.) rather than mocks.
  - Added a "What gets tested" section : the seven loopback groups run by test_framework_run_cases(), each subtest
  either run or reported as a categorized SKIP, never silently dropped.
  - Added build/flash instructions for the three interrupt modes (GIC / cascaded AXI-INTC / polled noirq), all built
  from one source with only the .xsa and build dir changing.
  - Added a "Which IP per build" grid keyed to the auto-detected IRQ mode, plus the basic UART-only smoke-test build.
  - Added Cora Z7-07S wiring (shared PMOD straps across all three builds, console on PS UART0) and STM32 NUCLEO-F767ZI
  wiring notes.

  test_irq.c

  - Dropped the edges / delivered / per_edge TEST_VALUE prints(not useful info unless debugging)

**Review by buha — DISMISSED** [review]

**Comment by buha** [comment]
need to find a way to download those .xsa files using build_projects.py so that CI can build

File: projects/capi_selftest/boards/basic/nucleo-f767zi.conf
Comment (rbolboac) [inline projects/capi_selftest/boards/basic/nucleo-f767zi.conf:1]: .ioc should be added here (see other projects for reference), otherwise the build will not work

**Comment by rbolboac** [comment]
change log: added CONFIG_XILINX_HDL_DESIGN="test_capi_selftest" and performed main rebase (xilinx should build faster)

File: projects/capi_selftest/boards/basic/nucleo-f767zi.conf [hunk 1]
```diff
@@ -0,0 +1 @@
+CONFIG_STM32_IOC_PATH="nucleo-f767zi.ioc"
```
Comment (rbolboac) [inline projects/capi_selftest/boards/basic/nucleo-f767zi.conf:1]: this change should be added in the commit which created the conf file and added the ioc, not in a new commit

**Review by rbolboac — APPROVED** [review]
lgtm, let's wait for xilinx build to finish before merging

**Comment by rbolboac** [comment]
xilinx build fail - unrelated to these changes
